### PR TITLE
REST Api implementation

### PR DESCRIPTION
**Pull request by DreamVan**
Add REST API implementation for Proxy repository
[Jira NEXUS-22138](https://issues.sonatype.org/browse/NEXUS-22138)

### CODE REVIEW

**Comment by DreamVan** [comment]
@DarthHater please check CLA issue. I have no ability to solve it

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2RepositoriesApiResourceIT.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+import org.ops4j.pax.exam.Configuration;
+import org.ops4j.pax.exam.Option;
+import org.sonatype.nexus.common.app.BaseUrlHolder;
+import org.sonatype.nexus.repository.Repository;
+import org.sonatype.nexus.repository.p2.internal.P2Format;
+import org.sonatype.nexus.repository.p2.internal.P2ITSupport;
+import org.sonatype.nexus.repository.rest.api.model.AbstractRepositoryApiRequest;
+import org.sonatype.nexus.repository.types.ProxyType;
+
+import static javax.ws.rs.core.Response.*;
```
Comment (proteus1121) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2RepositoriesApiResourceIT.java]: please avoid star imports

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2RepositoriesApiResourceIT.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    AbstractRepositoryApiRequest request = createProxyRequest(false);
+
+    Response response = put(getUpdateRepositoryPathUrl(ProxyType.NAME, PROXY_NAME), request);
+    assertEquals(response.getStatus(), Status.NO_CONTENT.getStatusCode());
+
+    Repository repository = repositoryManager.get(request.getName());
+    assertNotNull(repository);
+
+    assertThat(repository.getConfiguration().attributes("storage")
```
Comment (proteus1121) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2RepositoriesApiResourceIT.java]: StorageFacetConstants.STORAGE

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2RepositoriesApiResourceIT.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            .get("strictContentTypeValidation"),
```
Comment (proteus1121) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2RepositoriesApiResourceIT.java]: StorageFacetConstants.STRICT_CONTENT_TYPE_VALIDATION

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2ResourceITSupport.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  protected void setUnauthorizedUser() {
+    String randomRoleName = "role_" + UUID.randomUUID().toString();
+
+    Role role = securityRule.createRole(randomRoleName, UNRELATED_PRIVILEGE);
+    securityRule.createUser(randomRoleName, randomRoleName, role.getRoleId());
+    credentials = new UsernamePasswordCredentials(randomRoleName, randomRoleName);
+  }
+
+  protected void setBadCredentials() {
+    credentials = new UsernamePasswordCredentials("fake_user", "fake_user");
```
Comment (proteus1121) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2ResourceITSupport.java]: trivial: user_pass?

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2ResourceITSupport.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  }
+
+  protected String getCreateRepositoryPathUrl(final String type) {
+    return new StringJoiner("/")
```
Comment (proteus1121) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2ResourceITSupport.java]: P2PathUtils.DIVIDER

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/rest/P2ProxyRepositoriesApiResourceDoc.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+{
+  @ApiOperation("Create Helm proxy repository")
+  @ApiResponses(value = {
+      @ApiResponse(code = 201, message = REPOSITORY_CREATED),
+      @ApiResponse(code = 401, message = AUTHENTICATION_REQUIRED),
+      @ApiResponse(code = 403, message = INSUFFICIENT_PERMISSIONS)
+  })
+  Response createRepository(@Valid @NotNull final P2ProxyRepositoryApiRequest request);
+
+  @ApiOperation("Update Helm proxy repository")
```
Comment (proteus1121) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/rest/P2ProxyRepositoriesApiResourceDoc.java]: P2

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/rest/P2ProxyRepositoriesApiResourceDoc.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import static org.sonatype.nexus.rest.ApiDocConstants.API_REPOSITORY_MANAGEMENT;
+import static org.sonatype.nexus.rest.ApiDocConstants.AUTHENTICATION_REQUIRED;
+import static org.sonatype.nexus.rest.ApiDocConstants.INSUFFICIENT_PERMISSIONS;
+import static org.sonatype.nexus.rest.ApiDocConstants.REPOSITORY_CREATED;
+import static org.sonatype.nexus.rest.ApiDocConstants.REPOSITORY_UPDATED;
+
+@Api(value = API_REPOSITORY_MANAGEMENT)
+public interface P2ProxyRepositoriesApiResourceDoc
+{
+  @ApiOperation("Create Helm proxy repository")
```
Comment (proteus1121) [inline nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/rest/P2ProxyRepositoriesApiResourceDoc.java]: P2

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2ResourceITSupport.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+  public static final String FORMAT_VALUE = P2Format.NAME;
+
+  public static final String PROXY_NAME = String.format("%s-%s", FORMAT_VALUE, ProxyType.NAME);
+
+  // SET YOUR FORMAT DATA
+  public static final String UNRELATED_PRIVILEGE = "nx-analytics-all";
+
+  private static final String REPOSITORIES_API_URL = "service/rest/beta/repositories";
+
+  private static final ObjectMapper objectMapper = new ObjectMapper();
```
Comment (proteus1121) [inline nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2ResourceITSupport.java]: OBJECT_MAPPER

File: pom.xml [hunk 9]
```diff
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.sonatype.nexus.plugins</groupId>
     <artifactId>nexus-plugins</artifactId>
-    <version>3.19.0-01</version>
+    <version>3.20.0-SNAPSHOT</version>
   </parent>
```
Comment (mlukaretkyi) [inline pom.xml]: why you did not use the latest version?  The latest is 3.21 ?

Comment (DreamVan) [reply pom.xml]: It's not released and it should be a separate ticket

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2RepositoriesApiResourceIT.java (hunk 1 above)
Comment (DreamVan) [reply nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2RepositoriesApiResourceIT.java]: Done ae346d84d930c05248a3ffc4853de2819769f62e

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2RepositoriesApiResourceIT.java (hunk 2 above)
Comment (DreamVan) [reply nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2RepositoriesApiResourceIT.java]: Done ae346d84d930c05248a3ffc4853de2819769f62e

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2RepositoriesApiResourceIT.java (hunk 3 above)
Comment (DreamVan) [reply nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2RepositoriesApiResourceIT.java]: Done ae346d84d930c05248a3ffc4853de2819769f62e

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2ResourceITSupport.java (hunk 4 above)
Comment (DreamVan) [reply nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2ResourceITSupport.java]: Done ae346d84d930c05248a3ffc4853de2819769f62e

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2ResourceITSupport.java (hunk 5 above)
Comment (DreamVan) [reply nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2ResourceITSupport.java]: Done ae346d84d930c05248a3ffc4853de2819769f62e

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/rest/P2ProxyRepositoriesApiResourceDoc.java (hunk 6 above)
Comment (DreamVan) [reply nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/rest/P2ProxyRepositoriesApiResourceDoc.java]: Done ae346d84d930c05248a3ffc4853de2819769f62e

File: nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/rest/P2ProxyRepositoriesApiResourceDoc.java (hunk 7 above)
Comment (DreamVan) [reply nexus-repository-p2/src/main/java/org/sonatype/nexus/repository/p2/rest/P2ProxyRepositoriesApiResourceDoc.java]: Done ae346d84d930c05248a3ffc4853de2819769f62e

File: nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2ResourceITSupport.java (hunk 8 above)
Comment (DreamVan) [reply nexus-repository-p2-it/src/test/java/org/sonatype/nexus/repository/p2/rest/P2ResourceITSupport.java]: Done ae346d84d930c05248a3ffc4853de2819769f62e

**Pull request merged**